tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/Kinesis/KinesisContextAdapter.cs

```diff
@@ -25,36 +25,32 @@ public KinesisContextAdapter()
 
         public Dictionary<string, object> GetDictionary()
         {
-            var retval = new Dictionary<string, object>();
-            foreach (string key in headers.Keys)
-            {
-                retval.Add(key, headers[key][0]);
-            }
-
-            return retval;
+            // Convert to Dictionary<string, object> to satisfy IHeadersCollection
+            return headers.ToDictionary(kvp => kvp.Key, kvp => (object)kvp.Value);
```

Comment:
btw I was wondering how often is this called ? Would it make sense to directly have a dict of `<string, object>` and make the casts when we need to modify the list ? This would allow us to avoid making a copy of the dictionary here...

Reply:
The main path for this will be to create the adapter, add a small (<10) number of single entries, use it once, and destroy it.

The DSM manager is built around the expectation of some persistent headers that it can manipulate through the adapter interface, but because kinesis doesn't have headers or metadata this class is just wrapping a one-off that doesn't persist beyond one scope.

Reply:
Yeah, this could be a (small) performance issue if it's used in every Kinesis call. Not only is it allocating a new dictionary every time, but it's doing it using a LINQ call (`IEnumerable<KeyValuePair<string, List<string>>>.ToDictionary`) which will itself probably allocate an enumerator as well and cause the dictionary to grow as needed, which is wasteful.

If nothing else, I would at least remove the LINQ code and pre-initialize the new dictionary to the required size. Something like:
```csharp
var result = new Dictionary<string, object>(capacity: headers.Count);

foreach (var kvp in headers)
{
    result[kvp.Key] = kvp.Value;
}

return result;
```

It would be even better if the result was cacheable, or if it returned a sort of wrapper around `headers` instead of an entirely new object.